Propose new MaxPool-8 operation (#5359)
openvinotoolkit/openvino: docs/ops/pooling/MaxPool_8.md

**Short description**: Performs max pooling operation on input.

**Detailed description**: Input shape can be either 3D, 4D or 5D. Max Pooling operation is performed with the respect to input shape from the third dimension to the last dimension. If paddings are used then during the pooling calculation their value are `-inf`. The Max Pooling operation involves sliding a filter over each channel of feature map and downsampling by choosing the biggest value within the region covered by the filter. [Article about max pooling in Convolutional Networks](https://deeplizard.com/learn/video/ZjM_XQa5s6s). 

**Attributes**: *Pooling* attributes are specified in the `data` node, which is a child of the layer node.

* *strides*

  * **Description**: *strides* is a distance (in pixels) to slide the window on the feature map over the (z, y, x) axes for 3D poolings and (y, x) axes for 2D poolings. For example, *strides* equal "4,2,1" means sliding the window 4 pixel at a time over depth dimension, 2 over height dimension and 1 over width dimension.
  * **Range of values**: integer values starting from 0
  * **Type**: int[]
  * **Default value**: None
  * **Required**: *yes*

* *dilations*

>     * **Description**: *dilations* specify the index of the next pixel to select then pooling. If not present, the dilation defaults to 1, meaning the very next pixel is chosen. A value of 2 indicates that one pixel is skipped and every other pixel is considered. Dilations specify one value for each spatial axis of the kernel: `(z, y, x)` for 3D poolings and `(y, x)`  for 2D poolings.

Reviewer: ```suggestion
    * **Description**: *dilations* specify the index of the next pixel to select when pooling. If not present, the dilation defaults to 1, meaning the adjacent pixel is chosen. A value of 2 indicates that one pixel is skipped and every other pixel is considered. Dilations specify one value for each spatial axis of the kernel: `(z, y, x)` for 3D poolings and `(y, x)`  for 2D poolings.
```

## MaxPool <a name="MaxPool"></a> {#openvino_docs_ops_pooling_MaxPool_8}

**Versioned name**: *MaxPool-8*

**Category**: *Pooling*

**Short description**: Performs the max pooling operation on input.

> **Detailed description**: Input shape can be either 3D, 4D, or 5D. The max pooling operation is performed with respect to input shape from the third dimension to the last dimension. If paddings are used, during the pooling calculation their values are `-inf`. The max pooling operation involves sliding a filter over each channel of a feature map and downsampling by choosing the largest value within the region covered by the filter. [Article about max pooling in Convolutional Networks](https://deeplizard.com/learn/video/ZjM_XQa5s6s).

Reviewer 2: let's remove references to third-party sites.

  * **Range of values**: integer values starting from 0
  * **Type**: int[]
  * **Required**: *yes*

* *dilations*

    * **Description**: *dilations* specify the index of the next pixel to select when pooling. If not present, the dilation defaults to 1, meaning the adjacent pixel is chosen. A value of 2 indicates that one pixel is skipped and every other pixel is considered. Dilations specify one value for each spatial axis of the kernel: `(z, y, x)` for 3D poolings and `(y, x)`  for 2D poolings.
    * **Range of values**: integer values starting from 0
    * **Type**: int[]
    * **Default value**: `[1,1,...]`
    * **Required**: *no*

* *pads_begin*

  * **Description**: *pads_begin* is a number of pixels to add to the beginning along each axis. For example, *pads_begin* equal to "1,2" means adding 1 pixel to the top of the input and 2 to the left of the input.
  * **Range of values**: integer values starting from 0
  * **Type**: int[]
  * **Required**: *yes*
  * **Note**: the attribute is ignored when *auto_pad* attribute is specified.

* *pads_end*

>   * **Description**: *pads_end* is a number of pixels to add to the ending along each axis. For example, *pads_end* equal to "1,2" means adding 1 pixel to the bottom of the input and 2 to the right of the input.

Reviewer 2: May be add somewhere in the documentation that the padded values are equal to -inf?

**Attributes**: *Pooling* attributes are specified in the `data` node, which is a child of the layer node.

* *strides*

  * **Description**: *strides* is a distance (in pixels) to slide the window on the feature map over the (z, y, x) axes for 3D poolings and (y, x) axes for 2D poolings. For example, *strides* equal to "4,2,1" means sliding the window 4 pixels at a time over depth dimension, 2 over height dimension, and 1 over width dimension.
  * **Range of values**: integer values starting from 0
  * **Type**: int[]
  * **Required**: *yes*

* *dilations*

    * **Description**: *dilations* specify the index of the next pixel to select when pooling. If not present, the dilation defaults to 1, meaning the adjacent pixel is chosen. A value of 2 indicates that one pixel is skipped and every other pixel is considered. Dilations specify one value for each spatial axis of the kernel: `(z, y, x)` for 3D poolings and `(y, x)`  for 2D poolings.
    * **Range of values**: integer values starting from 0
    * **Type**: int[]
    * **Default value**: `[1,1,...]`
    * **Required**: *no*

> * *pads_value*

Reviewer: Suggestion to add a requirement that the value needs to fit in the data type of the operator's input. Otherwise the op might produce undefined behavior.

        `W_out = floor((W + pads_begin[1] + pads_end[1] - ((kernel[1] - 1) * dilations[1] + 1)) / strides[1] + 1)`
        `D_out = floor((D + pads_begin[2] + pads_end[2] - ((kernel[2] - 1) * dilations[2] + 1)) / strides[2] + 1)`

  * `auto_pad = explicit` and `rounding_type = ceil`
        `H_out = ceil((H + pads_begin[0] + pads_end[0] - ((kernel[0] - 1) * dilations[0] + 1)) / strides[0] + 1)`
        `W_out = ceil((W + pads_begin[1] + pads_end[1] - ((kernel[1] - 1) * dilations[1] + 1)) / strides[1] + 1)`
        `D_out = ceil((D + pads_begin[2] + pads_end[2] - ((kernel[2] - 1) * dilations[2] + 1)) / strides[2] + 1)`

  * `auto_pad = valid`
        `H_out = ceil((H - ((kernel[0] - 1) * dilations[0] + 1) + 1) / strides[0])`
        `W_out = ceil((W - ((kernel[1] - 1) * dilations[1] + 1) + 1) / strides[1])`
        `D_out = ceil((D - ((kernel[2] - 1) * dilations[2] + 1) + 1) / strides[2])`

  * `auto_pad = same_upper / same_lower`
        `H_out = H`
        `W_out = W`
        `D_out = D`

If `H + pads_begin[i] + pads_end[i] - kernel[i]` is not divisible by `strides[i]` evenly then the result is rounded with the respect to `rounding_type` attribute.

> Example 1 shows how *MaxPool* operates with 4D input using 2D kernel and `auto_pad = explicit`

Reviewer 3: ```suggestion
Example 1 shows how *MaxPool* operates with 4D input using 2D kernel and `auto_pad = explicit`.
```

```
input = [[[[-1, 2, 3],
           [4, 5, -6],
           [-7, 8, 9]]]]
strides = [1, 1]
pads_begin = [1, 1]
pads_end = [1, 1]
kernel = [2, 2]
rounding_type = "floor"
auto_pad = "explicit"
output0 = [[[[-1, 2, 3, 3],
             [4, 5, 5, -6],
             [4, 8, 9, 9],
             [-7, 8, 9, 9]]]]
output1 = [[[[0, 1, 2, 2],
             [3, 4, 4, 5],
             [3, 7, 8, 8],
             [6, 7, 8, 8]]]]
```

> Example 2 shows how *MaxPool* operates with 3D input using 1D kernel and `auto_pad = valid`

Reviewer 3: ```suggestion
Example 2 shows how *MaxPool* operates with 3D input using 1D kernel and `auto_pad = valid`.
```

```
input = [[[-1, 2, 3, 5, -7, 9, 1]]]
strides = [1]
kernel = [3]
rounding_type = "floor"
auto_pad = "valid"
output0 = [[[3, 5, 5, 9, 9]]]
output1 = [[[2, 3, 3, 5, 5]]]
```

> Example 3 shows how *MaxPool* operates with 4D input using 2D kernel and `auto_pad = same_lower`

Reviewer 3: ```suggestion
Example 3 shows how *MaxPool* operates with 4D input using 2D kernel and `auto_pad = same_lower`.
```

```
input = [[[[-1, 2, 3],
         [4, 5, -6],
         [-7, 8, 9]]]]
strides = [1, 1]
kernel = [2, 2]
rounding_type = "floor"
auto_pad = "same_lower"
output0 = [[[[-1, 2, 3],
            [4, 5, 5]
            [4, 8, 9]]]]
output1 = [[[[0, 1, 2],
            [3, 4, 4]
            [3, 7, 8]]]]
```

> Example 4 shows how *MaxPool* operates with 4D input using 2D kernel and `auto_pad = same_upper`

Reviewer 3: ```suggestion
Example 4 shows how *MaxPool* operates with 4D input using 2D kernel and `auto_pad = same_upper`.
```

```
input = [[[[-1, 2, 3],
           [4, 5, -6],
           [-7, 8, 9]],
          [[2, -1, 5],
           [6, -7, 1],
           [8, 2, -3]]]]
strides = [1, 1]
kernel = [2, 2]
rounding_type = "floor"
auto_pad = "same_upper"
output0 = [[[[5, 5, 3],
             [8, 9, 9]
             [8, 9, 9]],
            [[6, 5, 5],
             [8, 2, 1],
             [8, 2, -3]]]]
output1 = [[[[4, 4, 2],
             [7, 8, 8]
             [7, 8, 8]],
            [[12, 11, 11],
             [15, 16, 14],
             [15, 16, 17]]]]
```

> Example 5 shows how *MaxPool* operates with 4D input using 2D kernel, `auto_pad = valid` and `rounding_type = ceil`

Reviewer 3: ```suggestion
Example 5 shows how *MaxPool* operates with 4D input using 2D kernel, `auto_pad = valid` and `rounding_type = ceil`.
```

```
input = [[[[-1, 2, 3],
           [4, 5, -6],
           [-7, 8, 9]]]]
strides = [2, 2]
kernel = [2, 2]
rounding_type = "ceil"
auto_pad = "valid"
output0 = [[[[5, 3],
             [8, 9]]]]
output1 = [[[[4, 2],
             [7, 8]]]]
```

> Example 6 shows how *MaxPool* operates on 4D input using dilated 2D kernel, `auto_pad = explicit` and `rounding_type = floor`

Reviewer 3: ```suggestion
Example 6 shows how *MaxPool* operates on 4D input using dilated 2D kernel, `auto_pad = explicit` and `rounding_type = floor`.
```

```
input = [[[[1, 2, 3],
           [4, 5, 6],
           [7, 8, 9]]]]
strides = [1, 1]
kernel = [2, 2]
dilations = [2, 2]
rounding_type = "floor"
auto_pad = "explicit"
pads_begin = [1, 1]
pads_end = [1, 1]
output0 = [[[[5, 6, 5],
             [8, 9, 8],
             [5, 6, 5]]]]
output1 = [[[[4, 5, 4],
             [7, 8, 7],
             [4, 5, 4]]]]
```

> Example 7 shows how *MaxPool* operates on 4D input using 2D kernel, with non-default `axis` value

Reviewer 3: ```suggestion
Example 7 shows how *MaxPool* operates on 4D input using 2D kernel, with non-default `axis` value.
```

**Versioned name**: *MaxPool-8*

**Category**: *Pooling*

**Short description**: Performs max pooling operation on input.

**Detailed description**: Input shape can be either 3D, 4D or 5D. Max Pooling operation is performed with the respect to input shape from the third dimension to the last dimension. If paddings are used then during the pooling calculation their value are `-inf`. The Max Pooling operation involves sliding a filter over each channel of feature map and downsampling by choosing the biggest value within the region covered by the filter. [Article about max pooling in Convolutional Networks](https://deeplizard.com/learn/video/ZjM_XQa5s6s).

**Attributes**: *Pooling* attributes are specified in the `data` node, which is a child of the layer node.

* *strides*

  * **Description**: *strides* is a distance (in pixels) to slide the window on the feature map over the (z, y, x) axes for 3D poolings and (y, x) axes for 2D poolings. For example, *strides* equal "4,2,1" means sliding the window 4 pixel at a time over depth dimension, 2 over height dimension and 1 over width dimension.
  * **Range of values**: integer values starting from 0
  * **Type**: int[]

>   * **Default value**: None

Reviewer 2: We do not add "Default value" if the attribute is "Required" anymore.

  * **Required**: *yes*

* *dilations*

    * **Description**: *dilations* specify the index of the next pixel to select when pooling. If not present, the dilation defaults to 1, meaning the adjacent pixel is chosen. A value of 2 indicates that one pixel is skipped and every other pixel is considered. Dilations specify one value for each spatial axis of the kernel: `(z, y, x)` for 3D poolings and `(y, x)`  for 2D poolings.
    * **Range of values**: integer values starting from 0
    * **Type**: int[]

>     * **Default value**: [1,1,...]

Reviewer 2: wrap with "`" characters

  * **Range of values**:
    * *ceil*
    * *floor*
  * **Type**: string
  * **Default value**: *floor*
  * **Required**: *no*

* *auto_pad*

  * **Description**: *auto_pad* how the padding is calculated. Possible values:
    * *explicit*: use explicit padding values from `pads_begin` and `pads_end`.
    * *same_upper (same_lower)* the input is padded to match the output size. In case of odd padding value an extra padding is added at the end (at the beginning).
    * *valid* - do not use padding.
  * **Type**: string
  * **Default value**: *explicit*
  * **Required**: *no*
  * **Note**: *pads_begin* and *pads_end* attributes are ignored when *auto_pad* is not equal to explicit.

* *index_element_type*

    * **Description**: the type of output tensor with indices
    * **Range of values**: "i64" or "i32"
    * **Type**: string
    * **Default value**: "i64"
    * **Required**: *No*

* *axis*

    * **Description**: indicator of the first dimension in the input shape that should be used to calculate the upper bound of allowed index output values. The upper bound is the product of dimensions starting from the one pointed by the 'axis' attribute until the end of the input shape.

>     * **Range of values**: integer number. Negative value means counting dimension from the end. The range is `[-R, R-1]`, where `R` is the rank of the input tensor.

Reviewer 2: Change `[-R, R-1]` to `[-R, R - 1]`

    * **Type**: int
    * **Default value**: 0
    * **Required**: *No*

**Inputs**:

*   **1**: 3D, 4D or 5D input tensor of type T. Required.

**Outputs**:
  * **1**: Input shape can be either `[N, C, H]`, `[N, C, H, W]` or `[N, C, H, W, D]`. Then the corresponding output shape will be `[N, C, H_out]`, `[N, C, H_out, W_out]` or `[N, C, H_out, W_out, D_out]`. Output tensor has the same data type as input tensor.

  * **2**: Output tensor of type *T_IND* with indices of values selected by the pooling operation.
    Shape of this output matches the first output. The type of this output can be specified using the `index_element_type` attribute.

>     Values are computed as indices in a tensor flattened to 1-D, not considering padding. Examples for 5D input tensor:

Reviewer 2: We usually write `1D` instead of `1-D`, aren't we?

* *T_IND*: `int64` or `int32`.

**Mathematical Formulation**
Output shape calculation based on `auto_pad` and `rounding_type`:
  * `auto_pad = explicit` and `rounding_type = floor`
        `H_out = floor((H + pads_begin[0] + pads_end[0] - ((kernel[0] - 1) * dilations[0] + 1)) / strides[0] + 1)`
        `W_out = floor((W + pads_begin[1] + pads_end[1] - ((kernel[1] - 1) * dilations[1] + 1)) / strides[1] + 1)`
        `D_out = floor((D + pads_begin[2] + pads_end[2] - ((kernel[2] - 1) * dilations[2] + 1)) / strides[2] + 1)`

  * `auto_pad = explicit` and `rounding_type = ceil`
        `H_out = ceil((H + pads_begin[0] + pads_end[0] - ((kernel[0] - 1) * dilations[0] + 1)) / strides[0] + 1)`
        `W_out = ceil((W + pads_begin[1] + pads_end[1] - ((kernel[1] - 1) * dilations[1] + 1)) / strides[1] + 1)`
        `D_out = ceil((D + pads_begin[2] + pads_end[2] - ((kernel[2] - 1) * dilations[2] + 1)) / strides[2] + 1)`

  * `auto_pad = valid`
        `H_out = ceil((H - ((kernel[0] - 1) * dilations[0] + 1) + 1) / strides[0])`
        `W_out = ceil((W - ((kernel[1] - 1) * dilations[1] + 1) + 1) / strides[1])`
        `D_out = ceil((D - ((kernel[2] - 1) * dilations[2] + 1) + 1) / strides[2])`

>   * `auto_pad = same_upper/same_lower`

Reviewer 2: Add spaces around `/`

             [4, 5, 5, -6],
             [4, 8, 9, 9],
             [-7, 8, 9, 9]]]]
output1 = [[[[0, 1, 2, 2],
             [3, 4, 4, 5],
             [3, 7, 8, 8],
             [6, 7, 8, 8]]]]
```

Example 2 shows how *MaxPool* operates with 3D input using 1D kernel and `auto_pad = valid`

```
input = [[[-1, 2, 3, 5, -7, 9, 1]]]
strides = [1]
kernel = [3]
rounding_type = "floor"
auto_pad = "valid"
output0 = [[[3, 5, 5, 9, 9]]]
output1 = [[[2, 3, 3, 5, 5]]]
```

Example 3 shows how *MaxPool* operates with 4D input using 2D kernel and `auto_pad = same_lower`

```
input = [[[[-1, 2, 3],
         [4, 5, -6],
         [-7, 8, 9]]]]
strides = [1, 1]
kernel = [2, 2]
rounding_type = "floor"
auto_pad = "same_lower"
output0 = [[[[-1, 2, 3],
            [4, 5, 5]
            [4, 8, 9]]]]
output1 = [[[[0, 1, 2],
            [3, 4, 4]
            [3, 7, 8]]]]
```

Example 4 shows how *MaxPool* operates with 4D input using 2D kernel and `auto_pad = same_upper`

```
input = [[[[-1, 2, 3],
           [4, 5, -6],
           [-7, 8, 9]],
          [[2, -1, 5],
           [6, -7, 1],
           [8, 2, -3]]]]
strides = [1, 1]
kernel = [2, 2]
rounding_type = "floor"
auto_pad = "same_upper"

> output0 = [[[[5, 5, 3],

Reviewer 2: You have fixed a bug in the specification. Please, update the MaxPool_1.md as well.
